File: app/src/main/java/com/fivesysdev/standwithukraine/data/DayStatisticRepositoryImpl.java
package com.fivesysdev.standwithukraine.data;

import static java.time.temporal.TemporalAdjusters.firstDayOfMonth;

import android.util.Log;

import java.io.IOException;
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.FutureTask;

public final class DayStatisticRepositoryImpl implements DayStatisticRepository {

    List<DayStatistic> statistics;
    LocalDate fromDate;

    public DayStatisticRepositoryImpl() {
        statistics = new ArrayList<>();
        fromDate = LocalDate.now().with(firstDayOfMonth());
        updateStatistics();
    }

    private void updateStatistics() {
        Callable<List<DayStatistic>> task = () -> {
            Client client = new Client();
            try {
                return client.getStatisticsFromDate(fromDate.toString());
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        };
        FutureTask<List<DayStatistic>> future = new FutureTask<>(task);
        new Thread(future).start();
        try {
            addAll(future.get());
        } catch (InterruptedException | ExecutionException exception) {
            Log.d("REP", "Interrupted|ExecutionException");
        }
    }

    @Override
    public DayStatistic getStatisticByDate(String date) {
        if (LocalDate.parse(date).isBefore(fromDate)) {
            fromDate = fromDate.minusMonths(1);
            updateStatistics();
        }
        for (DayStatistic statistic: statistics) {
            if (statistic.getDate().equals(date)) {
                return statistic;
            }
        }
        return null;
    }

    private void addAll(List<DayStatistic> list) {
        for (DayStatistic item: list) {
            if (statistics.stream().noneMatch(element -> element.isEqual(item))) {
                statistics.add(item);
            }
        }
    }
}

File: app/src/main/java/com/fivesysdev/standwithukraine/data/DayStatisticRepositoryImpl.kt
package com.fivesysdev.standwithukraine.data

import java.time.LocalDate
import java.time.temporal.TemporalAdjusters
import java.util.concurrent.Callable
import java.util.concurrent.FutureTask

class DayStatisticRepositoryImpl : DayStatisticRepository {
    private var statistics: MutableList<DayStatistic> = ArrayList()
    private var fromDate: LocalDate = LocalDate.now().with(TemporalAdjusters.firstDayOfMonth())

    init {
        updateStatistics()
    }

    private fun updateStatistics() {
        val task = Callable {
            return@Callable StatisticService().getStatisticFromDate(fromDate.toString())
        }
        val future = FutureTask(task)
        Thread(future).start()
        addAll(future.get())
    }

    override fun getStatisticByDate(date: String?): DayStatistic? {
        val selectedDate = LocalDate.parse(date)
        if (selectedDate.isBefore(fromDate)) {
            fromDate = fromDate.minusMonths((fromDate.month - selectedDate.month.value.toLong()).value.toLong())
            updateStatistics()
        }
        var statistic = findStatistic(date)
        if (statistic == null) {
            fromDate = selectedDate
            updateStatistics()
            statistic = findStatistic(date)
        }
        return statistic
    }

    private fun findStatistic(date: String?) : DayStatistic? {
        for (statistic in statistics) {
            if (statistic.date == date) {
                return statistic
            }
        }
        return null
    }

    private fun addAll(list: List<DayStatistic>) {
        for (item in list) {
            if (statistics.stream().noneMatch { element: DayStatistic -> element == item }) {
                statistics.add(item)
            }
        }
    }
}
